### PR TITLE
CIP-0098? | Distributed Artifact Token Metadata

### PR DESCRIPTION
**Pull request by wout**
In the past one and a half years, I've released three on-chain generative token collections on Cardano, and I developed this standard out of necessity.

Introductory information is available here:
https://docs.venster.art/dats.html

And a slightly more detailed version of this CIP is available here:
https://docs.venster.art/dat-metadata-standard.html

The content of the CIP on our documentation page is the same as in this PR but with a bit more inline information.

------

[Rendered Proposal in Branch](https://github.com/venster-io/CIPs/blob/master/CIP-XXXX/README.md)

### CODE REVIEW

File: CIP-0098/README.md [hunk 1]
```diff
@@ -0,0 +1,474 @@
+---
+CIP: 98?
```
Comment (Ryun1) [inline CIP-0098/README.md:2]:
```suggestion
CIP: ?
```

Comment (Ryun1) [reply CIP-0098/README.md:2]: Editors will assign a number in a CIP Editors meeting 🤓

File: CIP-0098/README.md [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+Title: DAT Metadata Standard
```
Comment (Ryun1) [inline CIP-0098/README.md:3]:
```suggestion
Title: Distributed Artifact Token Metadata Standard
```
It's a good name but people finding this PR should more be able to quickly work out what its about without prior knowledge of the acronym.

File: CIP-0098/README.md [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+Category: Metadata
```
Comment (Ryun1) [inline CIP-0098/README.md:4]:
```suggestion
Category: Tokens
```
For me the category of `Tokens` fits this proposal a little better, as it relates to "About tokens (fungible or non-fungible) and minting policies in general." see [CIP-01 Categories](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0001#categories).

File: CIP-0098/README.md [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+Status: Proposed
+Authors:
+  - Wout Fierens <wout@venster.art>
+Discussions:
+  - https://github.com/cardano-foundation/cips/pulls/?
```
Comment (Ryun1) [inline CIP-0098/README.md:9]:
```suggestion
  - https://github.com/cardano-foundation/cips/pulls/530
```
Add PR link

File: CIP-0098/README.md [hunk 5]
```diff
@@ -0,0 +1,474 @@
...
+
+A single DAT is distributed over at least two tokens, represented by three possible token types: _scene_, _renderer_ and optionally, _dependency_.
+
+Typically, _scene_ tokens don't hold any code. They carry a reference to the _renderer_ token, arguments for invocation of the _renderer_ and token-specific properties. _Scene_ tokens are therefore relatively small (less than 3kB). All _scene_ tokens in a collection reference one, or at most a few, _renderer_ tokens.
+
+The code to render all _scene_ tokens is stored in a _renderer_ token, which in turn can reference _dependency_ tokens containing shared libraries or assets. _Dependency_ tokens are considered extensions to the _renderer_ and can not be referenced directly by _scene_ tokens.
+
+**Note**: The DAT Metadata Standard builds on the existing [CIP-0025](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0025) standard.
+
+## **1**. Scene
```
Comment (Ryun1) [inline CIP-0098/README.md:64]:
```suggestion
### **1**. Scene
```
Just to match CIP template.

File: CIP-0098/README.md [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+
+The _scene_ token is the part the end user will receive in their wallet. It contains all the information to render the DAT.
+
+### **1.a.** Metadata
```
Comment (Ryun1) [inline CIP-0098/README.md:68]:
```suggestion
#### **1.a.** Metadata
```

File: CIP-0098/README.md [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+Properties for the _scene_ token:
+
+- **`renderer`** (_required_): an object with two properties
+  - **`main`** (_required_): the `asset_name` of the _renderer_ token within the
+    current `policy_id` (e.g. `my_renderer`)
+  - **`arguments`** (_required_): an array with arbitrary values used as
+    arguments for the invocation of the _renderer_ (e.g.: `[123]`)
+- **`properties`** (_optional_): an object with arbitrary key/value pairs describing the token's (unique) properties
+
+### **1.b.** Argument directives
```
Comment (Ryun1) [inline CIP-0098/README.md:116]:
```suggestion
#### **1.b.** Argument directives
```

File: CIP-0098/README.md [hunk 8]
```diff
@@ -0,0 +1,474 @@
...
+  "@epoch.previous",
+  "@arguments.the_perfect_nft_000"
+]
+```
+
+Referencing another token's arguments does not work recursively.
+
+**Note**: The list of directives mentioned above is not final. Proposals for new directives will be considered and added over time.
+
+## **2**. Renderer
```
Comment (Ryun1) [inline CIP-0098/README.md:175]:
```suggestion
### **2**. Renderer
```

File: CIP-0098/README.md [hunk 9]
```diff
@@ -0,0 +1,474 @@
...
+
+The _renderer_ token is at the heart of a DAT collection and can either be a self-contained program or one with dependencies. It is always part of the same `policy_id` as the _scene_ tokens. Multiple _renderer_ tokens can exist within a policy, but _scene_ tokens can only reference one _renderer_ at a time.
+
+### **2.a.** Metadata
```
Comment (Ryun1) [inline CIP-0098/README.md:179]:
```suggestion
#### **2.a.** Metadata
```

File: CIP-0098/README.md [hunk 10]
```diff
@@ -0,0 +1,474 @@
...
+
+- **`outputType`** (_required_): the mime type of the renderer's output (it's up to the viewer to define the supported formats)
+- **`dependencies`** (_optional_): an array of objects with dependency definitions
+- **`browsers`** (_required for browser-based DATs_): an object with versions of browsers against which the _renderer_ is tested
+
+**Note**: While not mandatory, adding a **`license`** property to every file in the `files` section is advisable—more info on licenses is below in section 2.f.
+
+The _renderer_ token should be burned after minting to free up the UTxO.
+
+### **2.b.** On-chain dependencies
```
Comment (Ryun1) [inline CIP-0098/README.md:226]:
```suggestion
#### **2.b.** On-chain dependencies
```

File: CIP-0098/README.md [hunk 11]
```diff
@@ -0,0 +1,474 @@
...
+{
+  "type": "onchain",
+  "asset_name": <asset_name>
+}
+```
+
+- **`type`** (_required_): dependency type
+- **`asset_name`** (_required_): name of the asset within the same policy
+
+### **2.c.** Internal dependencies:
```
Comment (Ryun1) [inline CIP-0098/README.md:240]:
```suggestion
#### **2.c.** Internal dependencies:
```

File: CIP-0098/README.md [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+  "policy_id": <policy_id>,
+  "asset_name": <asset_name>
+}
+```
+
+- **`type`** (_required_): dependency type
+- **`asset_name`** (_required_): name of the asset
+- **`policy_id`** (_required_): policy id of the asset
+
+### **2.d.** External dependencies:
```
Comment (Ryun1) [inline CIP-0098/README.md:270]:
```suggestion
#### **2.d.** External dependencies:
```

File: CIP-0098/README.md [hunk 13]
```diff
@@ -0,0 +1,474 @@
...
+
+- **`type`** (_required_): dependency type
+- **`name`** (_required_): name of the library (list provided by viewers)
+- **`version`** (_required_): version of the library
+- **`source`** (_required_): URI of the library (IPFS, Arweave, etc.)
+- **`module`** (_required_): module flag; if set to `false`, an iife variant is used
+
+**Note**: The _external_ dependency definitions are not referencing any token on the blockchain, unlike the _onchain_ and _internal_ variants. Their sole purpose is instructing token viewers which external dependency to load at runtime. A list of available libraries is maintained in the [venster-external-dependencies repo](https://github.com/venster-io/venster-external-dependencies).
+
+### **2.e.** Build instructions
```
Comment (Ryun1) [inline CIP-0098/README.md:292]:
```suggestion
#### **2.e.** Build instructions
```

File: CIP-0098/README.md [hunk 14]
```diff
@@ -0,0 +1,474 @@
...
+{
+  "files": [{
+    "name": "Dockerfile",
+    "mediaType": "text/plain",
+    "src": <array>
+  }]
+}
+```
+
+### **2.f.** License types
```
Comment (Ryun1) [inline CIP-0098/README.md:308]:
```suggestion
#### **2.f.** License types
```

File: CIP-0098/README.md [hunk 15]
```diff
@@ -0,0 +1,474 @@
...
+For creators, it is recommended to choose a license that aligns with their values and with the purpose of the digital artifact. Any license can be used, but popular licenses are:
+
+- [NFT License 2.0](https://www.nftlicense.org/)
+- [CC BY-NC 4.0](https://creativecommons.org/licenses/by-nc/4.0/)
+- [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)
+- [AGPL 3.0](https://www.gnu.org/licenses/agpl-3.0.en.html)
+
+Using [no license](https://choosealicense.com/no-permission/) is also an option to explicitly indicate that no one other than the creator may use the software.
+
+## **3**. Dependency
```
Comment (Ryun1) [inline CIP-0098/README.md:319]:
```suggestion
### **3**. Dependency
```

File: CIP-0098/README.md [hunk 16]
```diff
@@ -0,0 +1,474 @@
...
+
+This section applies to _onchain_ and _internal_ dependencies. Aside from how they are referenced, they are technically the same.
+
+Dependency tokens can have multiple parts if they do not fit into one 16kB transaction. The dependency referenced from the _renderer_ serves as an entry point, referencing the additional parts.
+
+### **3.a.** Metadata
```
Comment (Ryun1) [inline CIP-0098/README.md:325]:
```suggestion
#### **3.a.** Metadata
```

File: CIP-0098/README.md [hunk 17]
```diff
@@ -0,0 +1,474 @@
...
+**Note**:
+While not mandatory, adding a **`license`** property to each file in the `files` section is advisable—more info on licenses is in section 2.f.
+
+Dependency tokens should be burned after minting to free up the UTxOs.
+
+## Rationale: how does this CIP achieve its goals?
+
+This section presents the resolutions to the four problems mentioned in the motivation section of this CIP. Given that this standard has already been embraced by a few collections, real-world examples are used to demonstrate the implementation and advantages of DATs.
+
+## Solution 1: **Storage limit**
```
Comment (Ryun1) [inline CIP-0098/README.md:364]:
```suggestion
### Solution 1: **Storage limit**
```

File: CIP-0098/README.md [hunk 18]
```diff
@@ -0,0 +1,474 @@
...
+
+The distributed nature of DATs allows bigger on-chain codebases with dependencies. The separation of code and metadata effectively doubles the available space.
+
+Additionally, dependencies can be stored in chunks, which allows for the realistic storage of up to fifty times more code than was previously possible.
+
+**Note**: It is important to note that the increase of storage capacity applies to _renderer_ tokens, not _scene_ tokens. Therefore, the additional space is only available for code used by the _renderer_ and not for static data like images in _scene_ tokens.
+
+### Example 1
```
Comment (Ryun1) [inline CIP-0098/README.md:372]:
```suggestion
#### Example 1
```

File: CIP-0098/README.md [hunk 19]
```diff
@@ -0,0 +1,474 @@
...
+| token type                 |   token size |   minted | collection size |
+| -------------------------- | -----------: | -------: | --------------: |
+| **scene**                  |      1.42 kB |     3409 |      4840.78 kB |
+| **renderer (JS/CSS/HTML)** |     12.85 kB |        1 |        12.85 kB |
+| **lib dependency (JS)**    |     11.77 kB |        1 |        11.77 kB |
+| **font dependency 1**      |     15.35 kB |        1 |        15.35 kB |
+| **font dependency 2**      |     15.24 kB |        1 |        15.24 kB |
+| **total**                  | **56.63 kB** | **3413** |  **4895.99 kB** |
+
+### Example 2
```
Comment (Ryun1) [inline CIP-0098/README.md:385]:
```suggestion
#### Example 2
```

File: CIP-0098/README.md [hunk 20]
```diff
@@ -0,0 +1,474 @@
...
+
+| token type                 |   token size |  minted | collection size |
+| -------------------------- | -----------: | ------: | --------------: |
+| **scene**                  |      2.44 kB |     512 |      1249.50 kB |
+| **renderer (JS/CSS/HTML)** |      9.72 kB |       1 |        12.85 kB |
+| **lib dependency (JS)**    |     11.77 kB |       1 |        11.77 kB |
+| **font dependency**        |     15.35 kB |       1 |        15.35 kB |
+| **total**                  | **39.28 kB** | **515** |  **1289.47 kB** |
+
+## Solution 2: **Inefficient use of block space**
```
Comment (Ryun1) [inline CIP-0098/README.md:397]:
```suggestion
### Solution 2: **Inefficient use of block space**
```

File: CIP-0098/README.md [hunk 21]
```diff
@@ -0,0 +1,474 @@
...
+
+By looking at three real-world cases, it becomes clear that using DATs requires roughly 90% less block space than using isolated monolithic NFTs.
+
+The data used in the following examples is extracted from existing token collections. The first two examples used monolithic NFTs, while the third one used DATs.
+
+### Example 1
```
Comment (Ryun1) [inline CIP-0098/README.md:403]:
```suggestion
#### Example 1
```

File: CIP-0098/README.md [hunk 22]
```diff
@@ -0,0 +1,474 @@
...
+This collection contains 17190 tokens as monolithic HTML NFTs.
+
+|                        | renderer size | token size | total kB | total MB |
+| ---------------------- | ------------: | ---------: | -------: | -------: |
+| **as monolithic NFTs** |           N/A |    5.34 kB | 91792 kB |  89.6 MB |
+| **as DATs**            |       4.91 kb |    0.53 kB |  9116 kB |   8.9 MB |
+
+This collection would have used **90.03 %** less block space with DATs
+
+### Example 2
```
Comment (Ryun1) [inline CIP-0098/README.md:414]:
```suggestion
#### Example 2
```

File: CIP-0098/README.md [hunk 23]
```diff
@@ -0,0 +1,474 @@
...
+This collection contains 1744 tokens as monolithic HTML NFTs.
+
+|                        | renderer size | token size | total kB | total MB |
+| ---------------------- | ------------: | ---------: | -------: | -------: |
+| **as monolithic NFTs** |           N/A |    13.9 kB | 24241 kB |  23.7 MB |
+| **as DATs**            |      13.40 kB |    1.46 kB |  2556 kB |   2.5 MB |
+
+This collection would have used **89.54 %** less block space with DATs.
+
+### Example 3
```
Comment (Ryun1) [inline CIP-0098/README.md:425]:
```suggestion
#### Example 3
```

File: CIP-0098/README.md [hunk 24]
```diff
@@ -0,0 +1,474 @@
...
+This is a collection of 3409 tokens and the first one ever minted using the DAT Metadata Standard.
+
+|                        | renderer size | token size |  total kB | total MB |
+| ---------------------- | ------------: | ---------: | --------: | -------: |
+| **as monolithic NFTs** |           N/A |   56.63 kB | 193052 kB | 188.5 MB |
+| **as DATs**            |      55.21 kB |    1.42 kB |   4896 kB |   4.8 MB |
+
+By using DATs, **97.46 %** less block space was used. It is also important to note that, since the token as monolithic NFT is larger than 16 kB, this collection wouldn't have been possible on Cardano.
+
+## Solutions 3 and 4: **Diversity of tools and archival qualities**
```
Comment (Ryun1) [inline CIP-0098/README.md:436]:
```suggestion
### Solutions 3 and 4: **Diversity of tools and archival qualities**
```

File: CIP-0098/README.md [hunk 25]
```diff
@@ -0,0 +1,474 @@
...
+
+Although problems 3 and 4 address different topics, one solution solves both: providing a standardized way to describe dependencies and build instructions.
+
+There are two variants of DATs: browser-based and non-browser-based. Almost all on-chain generative tokens on any blockchain fall into the first group. This standard supports both variants.
+
+### Variant 1: **Browser-based**
```
Comment (Ryun1) [inline CIP-0098/README.md:442]:
```suggestion
#### Variant 1: **Browser-based**
```

File: CIP-0098/README.md [hunk 26]
```diff
@@ -0,0 +1,474 @@
...
+
+Existing token viewers and blockchain explorers already support monolithic browser-based generative tokens. The most commonly used output formats are HTML and SVG, which can handle embedded JavaScript but can't have any dependencies.
+
+Through dependency definitions, DATs can instruct token viewers to load libraries, such as [p5.js](https://p5js.org/), to allow generative tokens beyond just plain JavaScript. Token viewers can decide which libraries, and versions thereof, to support.
+
+To enhance a token's archival qualities, DATs require a list of supported browsers with their versions in which the program is tested and confirmed to work correctly. It will help collectors to reproduce their token collections in the distant future.
+
+### Variant 2: **Non-browser-based**
```
Comment (Ryun1) [inline CIP-0098/README.md:450]:
```suggestion
#### Variant 2: **Non-browser-based**
```

File: CIP-0098/README.md [hunk 27]
```diff
@@ -0,0 +1,474 @@
...
+  "@block",
+  "@current_block",
+  "@epoch.previous",
+  "@arguments.the_perfect_nft_000"
+]
+```
+
+Referencing another token's arguments does not work recursively.
+
+**Note**: The list of directives mentioned above is not final. Proposals for new directives will be considered and added over time.
```
Comment (Ryun1) [inline CIP-0098/README.md:173]: What would be the process for considering new proposed amendments or updates to this specification?

**Review by Ryun1 — COMMENTED** [review]
Hey @wout, thanks for this interesting proposal!

Please see below the small _editorial_ type comments mainly about the structure of the proposal. And can you rename the directory from CIP-0098 to something like CIP-XXXX or CIP-????.

I do have one larger question:
- How would this proposal be versioned? since you mention this could be extended over time.

File: CIP-0098/README.md (hunk 1 above)
Comment (wout) [reply CIP-0098/README.md]: I understand 🙂️

File: CIP-0098/README.md (hunk 3 above)
Comment (wout) [reply CIP-0098/README.md:4]: Makes sense!

File: CIP-0098/README.md (hunk 5 above)
Comment (wout) [reply CIP-0098/README.md:64]: Right, I copied it straight from our docs, but this makes more sense indeed.

File: CIP-0098/README.md (hunk 27 above)
Comment (wout) [reply CIP-0098/README.md]: I'll come back on this in reply to your main comment at the top.

**Review by rphair — COMMENTED** [review]
@wout FYI we are introducing this proposal at the CIP editors' meeting today (4PM UTC).  I've seen you on Discord there; let me know if any other advocates of this proposal might be there & we can send them a Discord invite.

My preliminary review is:
- it seems to solve the problem that this deprecated proposal (https://github.com/cardano-foundation/CIPs/pull/249#issuecomment-1628994320) was trying to solve, without the features that led to its deprecation.  Therefore it could be instructive to readers to mention in the Rationale why this earlier considered alternative (adding an "external metadata" tag to existing metadata structures) was unpopular.
- That proposal was deprecated in favour of the draft CIP-0088 (https://github.com/cardano-foundation/CIPs/pull/467#issuecomment-1483920662), so could that other CIP approach be compared in the Rationale here as well?

You're not obligated to produce a CPS-like review of the whole problem, but to encourage people to adopt this as a standard it would be proactive to position it favourably compared to other potential solutions.

**Comment by rphair** [comment]
@wout I'm about to assign a number to this although the discussion at the meeting today was looking for more confirmation that this is used outside your own organisation.

Until that's established we have to remove the word `Standard` from the title at least... the feeling so far is that is a good proposal whose documentation can stand on its own in any case & could be merged as such if there's no counter-posting about why this wouldn't work.

Other editors were also interested in seeing a comparative explanation of your approach as requested in https://github.com/cardano-foundation/CIPs/pull/530#pullrequestreview-1556992846.

**Comment by rphair** [comment]
p.s. update to title requested in https://github.com/venster-io/CIPs/pull/1.

**Comment by rphair** [comment]
@wout I'm marking this `Waiting for Author` with no response to https://github.com/cardano-foundation/CIPs/pull/530#issuecomment-1660812529 after 5 weeks.

This PR came up in a review out outstanding proposals at the CIP Editors' meeting today and we noted some potential difficulties in passing review resulting from both these conditions:
- The proposal reflects particular design decisions & preferences which might not also be acceptable to a broader range of implementations.
- The implementation based on what you've posted as a CIP seems already to be in progress and therefore might not be possible to broaden after the fact.

Please respond about this reservation & what you might be prepared to do about it.  We still don't have all editors expressing opinions about this, but it might not be possible to approve this CIP if it's only documenting a single implementation.  If we don't have an answer with the usual review time frame we might have to eventually deprecate or close this.

**Comment by wout** [comment]
@rphair Sorry for the delayed reply. I'm going through some personal stuff and don't have the headspace for it at the moment.

**Comment by rphair** [comment]
Thanks for the update @wout ... feel free to open this again if & when you want to pursue the proposal further.

**Pull request closed without merging**